### ISSUE TITLE
miRsponge

### ISSUE THREAD

**Issue by zhangjunpeng411** [Bioconductor/Contributions#412]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/zhangjunpeng411/miRsponge

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by lshep** [comment]
Thank you for your submission to Bioconductor.  Some comments: 

From Build Report: 

 - [ ] You don't need to include the GPL LICENSE, listing it in the DESCRIPTION is enough.  
 - [ ] Add unit tests
 - [ ] Consider adding runnable examples to the following man pages which document exported objects:
      cernia.Rd, moduleFEA.Rd

README:
 - [ ] Add Bioconductor installation instructions in README via 
```
source("https://bioconductor.org/biocLite.R")
biocLite("miRsponge")
```

LICENSE: 
- [ ] as mentioned you don't need the copy of the GPL license
- [ ] You distribute based on GPL but reference http://www.apache.org/licenses/LICENSE-2.0 in your R code. Are these two license consistent with each other? Further more If you are utilizing someone else's code and are including the license for that reason this should be clearly acknowledged more readily in the DESCRIPTION file. 


R Code: 

- [ ] Currently there is no strong connection to Bioconductor objects. There needs to be a tighter correlation.  Your functions currently take in .csv files. Really they should take in R objects. 

miR2Target - should take a data.frame or look at GSEABase geneset object - maybe even provide a helper function to convert files from different technologies to the geneset. 
ExpData - should be a data.frame or summarizedExperiment 
MRE's - data.frame , maybe GSEABase geneset

- [ ] It is good practice to check the input of functions that the arguments are the expected class or formats

- [ ] It is currently really hard to follow code and suggest updates. You should adopt a consistent indent style so can follow loops. Please pass the R file into formatR that will assist with this. 


In general:
- [ ] don't use 1:n or 1:length(n) use seq_len or seq_along

- [ ] This appears several times:   
vectorize this function: 
 for (i in (1:num_perm)) {
        perm_seq[i, ] <- sample(num_sample)
  }	
In general check your loops to make sure vectorization isn't possible - sapply/vapply might possibly speed up efficiency as well. 

- [ ] You are setting seeds. This should not be necessary; the code should be general enough to not have to set a seed. 

- [ ] Repeated code is never encouraged. It seems like your main evaluation functions (miHomology, pcc, sppc, hermes, etc) do very similar things with some minor difference. At least a large part of the code is the same. They be combined into one larger function - consolodating the repeated code and have a
argument for the method (ie. method = c(pcc, hermes, ...)) and perform differences based on this flag? You could and probably should keep the indivdual functions for the methods but they would become wrappers to this larger functions.   

- [ ] integrateMethod, spongeValidate, netModule - all subset the results of the sponge interaction methods; This sub-setting should be done inside these functions so the user doesn't have to

- [ ] modelSurvival 
 expdatacsv should be an summarizedExperiment the summarizedExperiment can include the survival data in it, and have arguments to represent which columns contain that data. 

Please make updates and push changes.  When you are ready for a re-review please comment back here addressing the above concerns. Thank you

**Comment by zhangjunpeng411** [comment]
Thanks for your detail comments, @lshep. 
I will update the package ASAP.

Regards,
Junpeng

**Comment by zhangjunpeng411** [comment]
Hi @lshep,

Please find the below responses.

**From Build Report:**
1. GPL LICENSE has been removed.
2. The unit tests have been added.
3. The cernia function has been changed into an internal function. The examples in moduleFEA.Rd take much time during checking due to online functional enrichment analysis. So, I have pasted related codes but not run. 

**README:**
I have added Bioconductor installation instructions in README via
```
source("https://bioconductor.org/biocLite.R")
biocLite("miRsponge")
```

**LICENSE:**
1. The copy of the GPL license has been removed.
2. Internal functions of hermes and cernia method are from http://califano.c2b2.columbia.edu/hermes, Copyright Columbia University in the City of     New York. Internal functions of cernia method are from https://github.com/dsardina/cernia, Copyright 2016 Rosalba Giugno Licensed under the Apache License, Version 2.0. I have describe it directly in Description field of DESCRIPTION file.

**R Code:**
1. The input data of functions have been changed into data.frame object, including miRTarget, ExpData, mres, Groundtruth and SurvData.
2. I have checked that the arguments of functions are the expected class or formats.
3. The R file has been passed into formatR package to improve the format of R code.

**In general:**
1. 1:n or 1:length(n) has been changed using use seq_len or seq_along.
2. “ for (i in (1:num_perm)) { perm_seq[i, ] <- sample(num_sample)}” has been vectorized into “perm_seq <- t(sapply(seq_len(num_perm), function(i) sample(num_sample)))”. Some loops have also been vectorized or are replaced by sapply or vapply. Moreover, I also check some other loops and make sure that these loops cannot be vectorized. 
3. The seeds have been removed.
4. Thanks for the advice. I have changed miRHomology, pc, sppc, ppc, hermes, muTaME, and cernia into internal function. The spongeMethod function is created to consolidate the seven functions.
5. integrateMethod function can integrate any number of miRNA sponge interaction predictions methods. That is to say, the users can integrate 3 methods (e.g. miRHomology, pc and sppc) or 7 methods (e.g. miRHomology, pc, sppc, ppc, hermes, muTaME, and cernia) or k methods as they need. So, I reserve it separately.
spongeValidate function is used to validate predicted miRNA sponge interactions. 
netModule function is used identify miRNA sponge modules from miRNA sponge interaction network. The users may have their own methods to identify miRNA sponge interaction network. They can use this function to validate their predicted results, and identify miRNA sponge modules. So, I reserve them separately.
6. The input data (ExpData and SurvData) of moduleSurvival have been changed into data.frame objects.

Thanks again for your great advices and comments. 

Regards,
Junpeng

**Comment by zhangjunpeng411** [comment]
Hi  @lshep ,
There are more than two weeks with no updated comments for miRsponge package. I am afraid you miss my responses for the updated version. I attach the responses here again. Thank you for your review!

Hi  @lshep ,

Please find the below responses.

**From Build Report:**
1. GPL LICENSE has been removed.
2. The unit tests have been added.
3. The cernia function has been changed into an internal function. The examples in moduleFEA.Rd take much time during checking due to online functional enrichment analysis. So, I have pasted related codes but not run. 

**README:**
I have added Bioconductor installation instructions in README via
```
source("https://bioconductor.org/biocLite.R")
biocLite("miRsponge")
```

**LICENSE:**
1. The copy of the GPL license has been removed.
2. Internal functions of hermes and cernia method are from http://califano.c2b2.columbia.edu/hermes, Copyright Columbia University in the City of     New York. Internal functions of cernia method are from https://github.com/dsardina/cernia, Copyright 2016 Rosalba Giugno Licensed under the Apache License, Version 2.0. I have describe it directly in Description field of DESCRIPTION file.

**R Code:**
1. The input data of functions have been changed into data.frame object, including miRTarget, ExpData, mres, Groundtruth and SurvData.
2. I have checked that the arguments of functions are the expected class or formats.
3. The R file has been passed into formatR package to improve the format of R code.

**In general:**
1. 1:n or 1:length(n) has been changed using use seq_len or seq_along.
2. “ for (i in (1:num_perm)) { perm_seq[i, ] <- sample(num_sample)}” has been vectorized into “perm_seq <- t(sapply(seq_len(num_perm), function(i) sample(num_sample)))”. Some loops have also been vectorized or are replaced by sapply or vapply. Moreover, I also check some other loops and make sure that these loops cannot be vectorized. 
3. The seeds have been removed.
4. Thanks for the advice. I have changed miRHomology, pc, sppc, ppc, hermes, muTaME, and cernia into internal function. The spongeMethod function is created to consolidate the seven functions.
5. integrateMethod function can integrate any number of miRNA sponge interaction predictions methods. That is to say, the users can integrate 3 methods (e.g. miRHomology, pc and sppc) or 7 methods (e.g. miRHomology, pc, sppc, ppc, hermes, muTaME, and cernia) or k methods as they need. So, I reserve it separately.
spongeValidate function is used to validate predicted miRNA sponge interactions. 
netModule function is used identify miRNA sponge modules from miRNA sponge interaction network. The users may have their own methods to identify miRNA sponge interaction network. They can use this function to validate their predicted results, and identify miRNA sponge modules. So, I reserve them separately.
6. The input data (ExpData and SurvData) of moduleSurvival have been changed into data.frame objects.

Thanks again for your great advices and comments. 

Regards,
Junpeng

**Comment by lshep** [comment]
@zhangjunpeng411 I'm sorry for the late response.  I will re-evaluate the package and have more feedback either later today or tomorrow at the latest.

**Comment by zhangjunpeng411** [comment]
Thanks @lshep for your time.

Regards,
Junpeng

**Comment by zhangjunpeng411** [comment]
Thanks @lshep again for your time and comments.

Regards,
Junpeng

**Comment by lshep** [comment]
Thank you for your submission.  We appreciate your changes and think it helps make the package easier for the user and more integrated with other Bioconductor packages. Cheers

**Comment by zhangjunpeng411** [comment]
Hi @lshep @mtmorgan ,

There are more than two weeks with no additional information for the accepted _miRsponge_ package. Could I know when it will be added to the Bioconductor svn repository and nightly builds? Sorry for the bother. Thank you for your time!

Regards,
Junpeng

**Comment by lshep** [comment]
We apologize for the delay.  The package will be added soon; we have been finalizing the Git transition of Bioconductor packages before adding any additional new packages.

**Comment by zhangjunpeng411** [comment]
Hi @lshep ,

Thanks for your reply. I see.

Regards,
Junpeng

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to
Bioconductor's git repository.

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build
(check-out from git at about 6 pm Eastern; build completion around 2pm
Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main
landing page are consistent with the addition of your
package). Once the package builds successfully, you package will be
available for download in the 'Devel' version of Bioconductor using
`biocLite("YOUR_PACKAGE_NAME")`. The package 'landing page' will
be created at

https://bioconductor.org/packages/YOUR_PACKAGE_NAME

If you have any questions, please contact the bioc-devel mailing
list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue
will not be monitored further.
